ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayJoin.java

```diff
@@ -0,0 +1,111 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udf.array;
+
+import io.confluent.ksql.function.KsqlFunctionException;
+import io.confluent.ksql.function.udf.Udf;
+import io.confluent.ksql.function.udf.UdfDescription;
+import io.confluent.ksql.function.udf.UdfParameter;
+import io.confluent.ksql.util.KsqlConstants;
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.StringJoiner;
+import org.apache.kafka.connect.data.Struct;
+
+@SuppressWarnings("MethodMayBeStatic") // UDF methods can not be static.
+@UdfDescription(
+    name = "ARRAY_JOIN",
+    description = "joins the array elements into a flat string representation",
+    author = KsqlConstants.CONFLUENT_AUTHOR
+)
+public class ArrayJoin {
+
+  private static final String DEFAULT_DELIMITER = ",";
+  private static final Set<Class> KSQL_PRIMITIVES = new HashSet<>(
+      Arrays.asList(Boolean.class,Integer.class,BigInteger.class,Double.class,String.class)
+  );
+
+  @Udf
+  public <T> String join(
+      @UdfParameter(description = "the array to join using the default delimiter '"
+          + DEFAULT_DELIMITER + "'") final List<T> array
+  ) {
+    return join(array, DEFAULT_DELIMITER);
+  }
+
+  @Udf
+  public <T> String join(
+      @UdfParameter(description = "the array to join using the specified delimiter")
+      final List<T> array,
+      @UdfParameter(description = "the string to be used as element delimiter")
+      final String delimiter
+  ) {
+
+    if (array == null) {
+      return null;
+    }
+
+    final StringJoiner sj = new StringJoiner(delimiter);
+    array.forEach(e -> processElement(e, sj));
+    return sj.toString();
+
+  }
+
+  @SuppressWarnings("unchecked")
+  private static <T> void processElement(final T element, final StringJoiner joiner) {
+
+    if (element == null || KSQL_PRIMITIVES.contains(element.getClass())) {
+      handlePrimitiveType(element, joiner);
+    } else if (element instanceof List) {
+      handleListType((List)element,joiner);
+    } else if (element instanceof Map) {
+      handleMapType((Map) element, joiner);
+    } else if (element instanceof Struct) {
+      handleStructType((Struct)element, joiner);
+    } else {
+      throw new KsqlFunctionException("error: hit element of type "
+          + element.getClass().getTypeName() + " which is currently not supported");
+    }
+
+  }
+
+  private static void handlePrimitiveType(final Object element, final StringJoiner joiner) {
+    joiner.add(element != null ? element.toString() : null);
+  }
+
+  private static void handleListType(final List<?> element, final StringJoiner joiner) {
+    element.forEach(e -> processElement(e, joiner));
```

Comment:
if we have a list of list, should it flatten the nested types or just return a string representation of the element? it seems odd to me that `ARRAY[ARRAY[1, 2], ARRAY[3]]` would flatten to the same result as `ARRAY[1, 2, 3]`

same concern for map and struct - I think my preference would be to support only primitives in the first implementation. Other languages have similar restrictions, e.g. python (which is even more restrictive):
```python
>>> '.'.join([['1', '2']])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: sequence item 0: expected string, list found
```

Reply:
There were no clear indications about this in the issue discussion which is why I implemented it in this way, which is IMHO a "natural" choice. Remember that the original idea was that the result is just a flat string. If you want to see a different string result for the example you are referring to it essentially means to introduce further delimiters which can capture the semantics of nesting which might be out of scope for such a function. If someone really wants to have that I think it's better to decide for other forms of encoding e.g. a fully-fledged JSON-string representation, but I guess that's a different story.

Reply:
cc @derekjn @rmoff to chime in here, I'm happy with whatever we settle on

Reply:
I don't think we should implicitly flatten anything here. If I passed an array of arrays, I'd expect the result to be the string representation of each array, joined together by the delimiter. If that adds too much complexity to the implementation then I agree with @agavra that it would be reasonable to require primitive types, at least for the initial implementation. The semantics around concatenating string representations of structured types (especially if they have nondeterministic string reps) can start to get pretty messy.

Reply:
hi @derekjn so do I get you right that you want to capture the nesting semantics somehow in the resulting string representation? if so how _exactly_ should this look like? it would be helpful if you can provide to or some examples at least one for arrays of arrays, maps of maps or nested structs.

based on that we can then decide if the implementation should be done or we simply drop support for anything but flat arrays containing primitives. what I still quite don't get is why you have your issues with the current way it's implemented. think about a simple CSV-like string - in a basic CSV encoding you would end up with delimiting elements as is currently done since you don't capture complex structure in a simple CSV-style string.

Reply:
For all of those examples I think we'd just want to use whatever `CAST(value AS STRING)` would return for the array element. That being said, I'm fine with only supporting primitive types for the initial implementation as @agavra suggested. Let's just go with that.

Reply:
dropped support for complex nested types for this initial version of ARRAY_JOIN